### PR TITLE
Implemented 2 backend features

### PR DESCRIPTION
**Pull request by Krechals**
Features added:
1) Random 'Welcome' text using fetch() method.
2) Added logs.html page, which displays when the website was accessed.
Preview link: https://topandrei-step-2020.ew.r.appspot.com.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 1]
```diff
@@ -14,19 +14,35 @@
...
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
-@WebServlet("/data")
+@WebServlet("/random-welcome")
 public class DataServlet extends HttpServlet {
 
+  private List<String> welcomes;
+
+  @Override
+  public void init() {
+    welcomes = new ArrayList<>();
```
Comment (jeadorf) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: I recommend inlining the definition and making the member `final`.

`private List<String> welcomes = new ArrayList<>();
`
This is a simple mechanism of guaranteeing that this member will never be `null`.

File: portfolio/src/main/java/com/google/sps/servlets/ServerDatesServlet.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import java.io.IOException;
+import java.util.Date;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/server-dates")
+public final class ServerDatesServlet extends HttpServlet {
+  private List<Date> dateList = new ArrayList<>();
```
Comment (jeadorf) [inline portfolio/src/main/java/com/google/sps/servlets/ServerDatesServlet.java]: recommend: `final`

File: portfolio/src/main/java/com/google/sps/servlets/ServerDatesServlet.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    // Send the JSON as the response
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  /**
+   * Converts list of dates into a JSON string using the Gson library.
+   */
+  private String convertToJson(List<Date> dateList) {
```
Comment (jeadorf) [inline portfolio/src/main/java/com/google/sps/servlets/ServerDatesServlet.java]: This method can be `static`. This tells the reader that no state in the servlet instance is modified when invoking the method.

File: portfolio/src/main/java/com/google/sps/servlets/ServerDatesServlet.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/server-dates")
+public final class ServerDatesServlet extends HttpServlet {
+  private List<Date> dateList = new ArrayList<>();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Add another entry to the list
+    Date currentTime = new Date();
```
Comment (jeadorf) [inline portfolio/src/main/java/com/google/sps/servlets/ServerDatesServlet.java]: Consider looking into `java.time`.

**Review by jeadorf — CHANGES_REQUESTED** [review]
Generic note: whenever I see dates, I wonder about timezone and whether the timezone logic is working out as intended.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -14,19 +14,34 @@
...
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
-@WebServlet("/data")
+@WebServlet("/random-welcome")
 public class DataServlet extends HttpServlet {
 
+  private final List<String> welcomes = new ArrayList<>();;
```
Comment (jeadorf) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit: duplicate semicolon

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 6]
```diff
@@ -14,19 +14,34 @@
...
+
+  @Override
+  public void init() {
+    welcomes.add("Welcome!");
+    welcomes.add("Have a nice day!");
+    welcomes.add("Discover me!");
+    welcomes.add("Enjoy your day!");
+    welcomes.add("Welcome on board!");
+    welcomes.add("Fasten your seatbelts, we are ready for takeoff!");
+  }
```
Comment (jeadorf) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:39]: missing blank line?

https://google.github.io/styleguide/javaguide.html#s4.6.1-vertical-whitespace

File: portfolio/src/main/java/com/google/sps/servlets/ServerDatesServlet.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
+package com.google.sps.servlets;
+
+import java.util.ArrayList;
```
Comment (jeadorf) [inline portfolio/src/main/java/com/google/sps/servlets/ServerDatesServlet.java]: I know it's tedious (and you might find some tooling for it) but

https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing

File: portfolio/src/main/java/com/google/sps/servlets/ServerDatesServlet.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  /**
+   * Converts list of dates into a JSON string using the Gson library.
+   */
+  private static String convertToJson(List<ServerStats> dateList) {
+    Gson gson = new Gson();
+    String json = gson.toJson(dateList);
```
Comment (jeadorf) [inline portfolio/src/main/java/com/google/sps/servlets/ServerDatesServlet.java]: very minor nit: you could consider dropping the `json` variable and just write `return gson.toJson(dateList)`, which I already find very expressive.

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -12,6 +12,34 @@
...
 
+/**
+ * Adds a random welcome to the page.
+ */
+async function getRandomWelcome() {
+  const response = await fetch('/random-welcome');
+  const welcome = await response.text();
+  document.getElementById('welcome-msg').innerText = welcome;
+}
+async function getServerDates() {
```
Comment (jeadorf) [inline portfolio/src/main/webapp/script.js:24]: I think https://google.github.io/styleguide/jsguide.html#formatting-vertical-whitespace suggests an empty line between functions

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -12,6 +12,34 @@
...
+  const response = await fetch('/server-dates');
+  const dates = await response.json();
+  const body = document.getElementsByTagName('body')[0];
+  body.innerHTML = '';
+  dates.forEach(function(stats) {
```
Comment (jeadorf) [inline portfolio/src/main/webapp/script.js:29]: Note: I learnt about arrow functions today `stats => {}`. But your code will work, too.

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -12,6 +12,34 @@
...
+    body.appendChild(
+        createListElement('Server was accessed on: ' + stats.date.day + '/'
```
Comment (jeadorf) [inline portfolio/src/main/webapp/script.js:31]: This is fine. I'm wondering whether the backtick syntax would also work here

`Server was accessed on: ${stats.date.day}/${stats.date.month}/${stats.date.year} ...`

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -12,6 +12,34 @@
...
+  const response = await fetch('/server-dates');
+  const dates = await response.json();
+  const body = document.getElementsByTagName('body')[0];
+  body.innerHTML = '';
+  dates.forEach(function(stats) {
+    body.appendChild(
+        createListElement('Server was accessed on: ' + stats.date.day + '/' 
+                                                     + stats.date.month + '/' 
+                                                     + stats.date.year + ' at ' 
+                                                     + stats.time.minute + ':'
```
Comment (jeadorf) [inline portfolio/src/main/webapp/script.js]: hh:mm?

**Review by jeadorf — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 11 above)
Comment (Krechals) [reply portfolio/src/main/webapp/script.js:31]: I get a syntax error

**Review by jeadorf — APPROVED** [review]

**Pull request merged**